rct229/rulesets/ashrae9012019/section19/section19rule21.py

```diff
@@ -0,0 +1,498 @@
+                    else:
+                        zone_data[hvac_id_b][
+                            "all_ventilation_space_types_defined"
+                        ] = False
+
+                    for hvac_id_p in get_list_hvac_systems_associated_with_zone(
+                        rmd_p, zone_id_b
+                    ):
+                        hvac_p = find_one(
+                            f'$.buildings[*].building_segments[*].heating_ventilating_air_conditioning_systems[*][?(@.id = "{hvac_id_p}")]',
+                            rmd_p,
+                        )
+                        ER_modeled_p = find_one(
+                            "$.fan_system.air_energy_recovery.energy_recovery_type",
+                            hvac_p,
+                        ) in (
+                            ENERGY_RECOVERY.SENSIBLE_HEAT_EXHANGE,
+                            ENERGY_RECOVERY.ENTHALPY_HEAT_EXHANGE,
+                            ENERGY_RECOVERY.SENSIBLE_HEAT_WHEEL,
+                            ENERGY_RECOVERY.ENTHALPY_HEAT_WHEEL,
+                            ENERGY_RECOVERY.HEAT_PIPE,
+                            ENERGY_RECOVERY.OTHER,
+                        )
+
+                        serves_zones_that_have_dehumid_heat_recovery = (
+                            find_one("$.cooling_system.dehumidification_type", hvac_p)
+                            == DEHUMIDIFICATION.SERIES_HEAT_RECOVERY
+                        )
+
+                    hvac_systems_and_assoc_zones_largest_exhaust_source[
+                        hvac_id_b
+                    ] = get_hvac_sys_and_assoc_zones_largest_exhaust_source(
+                        rmd_b, hvac_id_b
+                    )
+
+                sys_type_heating_only = any(
+                    [
+                        hvac_id_b in baseline_hvac_system_dict[applicable_sys_type]
+                        for applicable_sys_type in APPLICABLE_SYS_TYPES
+                    ]
+                )
+
+            try:
+                max_thermostat_heating_setpoint_schedule_p = max(
+                    [
+                        max(thermostat_schedule)
+                        * ureg(
+                            "degC"
+                        )  # TODO: make sure this unit is degC. The current schema (v 0.0.29) doesn't specify the unit specifically
+                        for thermostat_schedule_id in find_all(
+                            "$.buildings[*].building_segments[*].zones[*].thermostat_heating_setpoint_schedule",
+                            rmd_p,
+                        )
+                        for thermostat_schedule in find_all(
+                            f'$.schedules[*][?(@.id = "{thermostat_schedule_id}")].hourly_values',
+                            rmd_p,
+                        )
+                    ]
+                )
+            except ValueError:
+                max_thermostat_heating_setpoint_schedule_p = ZERO.TEMPERATURE
```

Comment:
I don't think we need try and catch here.
1. I still think the serves_zones_heated_to_60_or_higher_p should be part of zone_data dictionary, not a stand-alone flag.
2. The list of zones shall be the `dict_of_zones_and_terminal_units_served_by_hvac_sys_b[hvac_id_b]["zone_list"]`
3. We can have a better way to handle this flag.
```python
thermostat_heating_setpoint_schedule_list = map_(
dict_of_zones_and_terminal_units_served_by_hvac_sys_b[hvac_id_b]["zone_list"], 
lambda zone_id: find_one("$.buildings[*].building_segments[*].zones[*][?(@.id = "{zone_id}")].thermostat_heating_setpoint_schedule", rmd_p)

# use this to remove duplications
thermostat_heating_setpoint_schedule_list = list(set(thermostat_heating_setpoint_schedule_list ))
# this flattens the 2d array to 1d array
thermostat_heating_setpoint_list = flat_map(thermostat_heating_setpoint_schedule_list , lambda schedule_id: find_exactly_one_schedule(rmd_p, schedule_id).get("hourly_value", [0]))

max_thermostat_heating_setpoint_schedule_p = max(thermostat_heating_setpoint_list) * ureg("degC")
```

Reply:
Addressed.